### ISSUE TITLE
Hindi font on pdf not working

### ISSUE THREAD

**Issue by rajn** [eclipse/birt#890]
I'm able to start using the birt designer after this #881 
The reason I wanted to try this new release was in hope of making Hindi fonts working in Birt.
I had seen some posts that suggested the use of itext by Birt was the root cause. I saw itext is now replaced in Birt 4.9.0.
But Hindi font is not working even with the new version of Birt.

**env**
----
OS - mac mojave
Birt - birt-report-designer-all-in-one-4.9.0-20220315-macosx.cocoa.x86_64

**Steps**
------
Create a blank report with a single label.
The label's text is loaded from a resource file.
Run the report. HTML shows the Hindi text  correctly
Change format to pdf in the url or export the report as pdf.
Now the report is empty without any Hindi text in it.

Complete set of files given below

**new_report.rptdesign**
-------
```
<?xml version="1.0" encoding="UTF-8"?>
<report xmlns="http://www.eclipse.org/birt/2005/design" version="3.2.24" id="1">
    <property name="createdBy">Eclipse BIRT Designer Version 4.9.0.v202203150031</property>
    <simple-property-list name="includeResource">
        <value>Resources_hi_IN</value>
        <value>Resources</value>
    </simple-property-list>
    <property name="units">in</property>
    <property name="iconFile">/templates/blank_report.gif</property>
    <property name="bidiLayoutOrientation">ltr</property>
    <property name="imageDPI">72</property>
    <page-setup>
        <simple-master-page name="Simple MasterPage" id="2">
            <page-footer>
                <text id="3">
                    <property name="contentType">html</property>
                    <text-property name="content"><![CDATA[<value-of>new Date()</value-of>]]></text-property>
                </text>
            </page-footer>
        </simple-master-page>
    </page-setup>
    <body>
        <label id="4">
            <text-property name="text" key="sincerely"></text-property>
        </label>
    </body>
</report>
```

**Resources_hi_IN.properties**
---------------------------
```
sincerely=\u0938\u093f\u0928\u094d\u0938\u093f\u0930\u094d\u0932\u0940
```

**Resources.properties**
```
sincerely=Sincerely
```

**Comment by claesrosell** [comment]
This maybe due to missing glyphs in the selected font. To reproduce this we need to know which font that you are targeting. I do not see any style information in the above .rptdesign

**Comment by rajn** [comment]
thanks @claesrosell. I'm not an expert in this area. any font is fine as long as it looks Hindi! For example I see DevanagariMT on the mac. I downloaded NotoSansDevanagari-Regular from google. If I can make this work with any such font i'm fine.
Can you please elaborate on this "I do not see any style information in the above .rptdesign". What should I do?

**Comment by claesrosell** [comment]
Try to explicitly select the DevanagariMT font on the label. It will not necessarily work when you have done that but at least you have tried.

I am a bit busy currently, I will not be able to help you any time soon. But if any of us do find time, It is nice ro have enough information to actually reproduce the problem.

**Comment by rajn** [comment]
thanks @claesrosell . I made the font on the label to be Devanagari MT but the result is still same. Hindi shows on HTML but on the PDF it is blank.

**rptdesign file**

```
<?xml version="1.0" encoding="UTF-8"?>
<report xmlns="http://www.eclipse.org/birt/2005/design" version="3.2.24" id="1">
    <property name="createdBy">Eclipse BIRT Designer Version 4.9.0.v202203150031</property>
    <simple-property-list name="includeResource">
        <value>Resources_hi_IN</value>
        <value>Resources</value>
    </simple-property-list>
    <property name="units">in</property>
    <property name="iconFile">/templates/blank_report.gif</property>
    <property name="bidiLayoutOrientation">ltr</property>
    <property name="imageDPI">72</property>
    <page-setup>
        <simple-master-page name="Simple MasterPage" id="2">
            <page-footer>
                <text id="3">
                    <property name="contentType">html</property>
                    <text-property name="content"><![CDATA[<value-of>new Date()</value-of>]]></text-property>
                </text>
            </page-footer>
        </simple-master-page>
    </page-setup>
    <body>
        <label id="4">
            <property name="fontFamily">"Devanagari MT"</property>
            <text-property name="text" key="sincerely"></text-property>
        </label>
    </body>
</report>
```

**Comment by hvbtup** [comment]
I don't know for sure, but I think the font handling with OpenPDF is more or less the same as with iText 2.1.7.

@rajn In order to check this, you can try to create a simple Java program that creates a PDF with Hindi text using the OpenPDF library (see https://github.com/LibrePDF/OpenPDF). There is an example at https://github.com/LibrePDF/OpenPDF/wiki/Multi-byte-character-language-support-with-TTF-fonts. If you can make that work, there is at least a chance that we can make it work with BIRT, too.

**Comment by speckyspooky** [comment]
Hello rajn,
I have testes your case with BIRT 4.8 and BIRT 4.9 on my windows-laptop.
The result is that on default BIRT couldn't find the font for PDF. There was a special configuration which I must set and afterwards the PDF-creation was possible with embedded fonts.

Configuration:
1. go to: C:\BIRT_4.9\eclipse\plugins\org.eclipse.birt.report.engine.fonts_4.9.0.v202203150031
2. change file: fontsConfig.xml   e.g. (my case)

&lt;font-paths>
    &lt;path path="C:/Users/gutma/Downloads/651_devanagari/Hindi-DVSR0ntt.ttf"/>
    &lt;path path="C:/windows/fonts" />
    ...
&lt;/font-paths>

After restart of BIRT the configuration will be catched.

For example my test case and the according PDF-document like attachment.
Try it on your side please.
![BIRT_4 9_Hindi_Font](https://user-images.githubusercontent.com/41593722/159789916-5c200ad0-8638-45f7-90e1-bbac23bb4f7b.png)
[Hindi.pdf](https://github.com/eclipse/birt/files/8336379/Hindi.pdf)

**Comment by hvbtup** [comment]
Ah yes. So the reason might have been the following:
You have installed that font on your Mac (note: I don't know anything about the apple parallel universe), but *the TrueType font file landed at some place where the default font configuration does not look*.

The font configuration files are a little bit complicated (but less complicated than those e.g. for Oracle Reports).
AFAIK there are 4 of them:

-  fontsConfig.xml
-  fontsConfig_linux.xml
-  fontsConfig_pdf.xml
-  fontsConfig_win32.xml

... and which of these files are actually used can be confusing.
Note that there is no specialized file for Mac.

You can simplify things a little bit by trying to understand what goes on in a single file, backup the files to a different folder, keep and adapt the file fontsConfig.xml and remove the other 3 files.

A `path` element in `font-paths` can be a single TTF file or a whole directory, e.g.

		<path path="/usr/X11R6/lib/X11/fonts/TTF" />

In case of a directory, BIRT will scan the directory for *.ttf files.

I had similar issues on newer Windows versions.
In the good old days, when you installed a font in Windows, it was copied to `%WINDIR%\fonts`.
However, things are more complicated nowadays.
I prefer to deploy a directory `fonts` with the ttf files in a known location and adding that known directory as a `path` in `fontsConfig.xml`.

**Comment by rajn** [comment]
Thanks a lot for your suggestions.. I can swear (who is checking : ) ) I had tried this with the old versions of Birt and it didnt work on Mac or Linux types OS. Windows always used to work for me but I had to make it work in Linux (production env). So now I tried the config xml change as suggested by the fine folks above and I can say it works on Mac. Now I need to try out on the actual application and I'm confident it will work. The change was to fontsConfig.xml where I added the location of the noto sans devanagari font ttf file like <path path="~/Library/Fonts/NotoSansDevanagari-Regular.ttf" /> and it works. 
Very happy to see Birt's rebirth.
